File: fourier.js
const drawCanvas = document.getElementById('drawCanvas');
const ctx = drawCanvas.getContext('2d');
const CX = 200, CY = 200, R = 120;



let N = 10; //number of points on the circle
let harmonics = N/2;
let points = [];
let playing = true;

//sliders and buttons ////////////////////////////
const harmonicsSlider = document.getElementById('hSlider');
harmonicsSlider.min = 1; harmonicsSlider.max = N; harmonicsSlider.value = N/2;
harmonicsSlider.addEventListener('input', () => {
    harmonics = parseInt(harmonicsSlider.value);
    document.getElementById('hVal').innerHTML = harmonics;
});
const nSlider = document.getElementById('nSlider');
nSlider.min = 1; nSlider.max = 20; nSlider.value = nSlider.max/2;
nSlider.addEventListener('input', () => {
    N = parseInt(nSlider.value);
    document.getElementById('nVal').innerHTML = N;
    initCircle();
    trace = []; // clear
    harmonicsSlider.max = N; harmonicsSlider.value = N/2;
    recompute();
});
//reset button
const rst = document.getElementById('resetButton');
rst.addEventListener('click', () => {
    initCircle();
    drawCircle();
});
//pause button
document.getElementById('pauseButton').addEventListener('click', (e) => {
    playing = !playing;
    e.target.textContent = playing ? "Pause" : "Play";
});
//sliders and buttons ////////////////////////////

// initalizes the location of points on the circle
function initCircle(){
    points = [];
    for(let i = 0; i < N; i++){
        const angle = (i/N) * 2 * Math.PI;
        points.push({x : Math.cos(angle) * R, y : Math.sin(angle) * R});
    }
}
initCircle();


//Discrete Fourier Transfrom! not the fast one. 
/* 
idea is: get the point, imagine it's in an imaginary plane, and then get the angle and the radius of that point. 
*/
function computeFourier(){
    const n = points.length;
    const coeffs = [];
    for(let k =0; k <n; k++){  //how much frequency we want to compute? 
        let re = 0, im = 0;
        for(let m = 0; m < n; m++) { //which point we're looking at
            const angle = -2 * Math.PI * k * m/n;
            const cos = Math.cos(angle);
            const sin = Math.sin(angle);
            re += points[m].x * cos - points[m].y * sin;
            im += points[m].x * sin + points[m].y * cos;
        }
        re /= n;
        im /=n;
        const freq = k <= n/2 ? k : k-n; //negative frequencies. 
        coeffs.push({freq, re, im});
    }
    coeffs.sort((a,b) => Math.abs(a.freq) - Math.abs(b.freq)); //sort by frequency
    console.log(coeffs);
    return coeffs;
}

function evalAt(t, terms){
    let x = CX, y = CY; //centre 
    const chain = [{x,y}];
    terms.forEach((term) => {
        const angle = term.freq * t;
        x += term.re * Math.cos(angle) - term.im * Math.sin(angle);
        y += term.re * Math.sin(angle) + term.im * Math.cos(angle); //rotation
        chain.push({x,y});
    });
    return chain; // last point in this chain is the final point of the fourier series at time t ! 
}

//reconstruction of the fourier series (right panel)
const display = document.getElementById('displayCanvas');
const dctx = display.getContext('2d');

let t = 0; trace = [];
function drawRight(coeffs){
    dctx.clearRect(0, 0, display.width, display.height);
    const terms = coeffs.slice(0, harmonics);
    const chain = evalAt(t, terms);
    for(let i = 0; i < chain.length -1; i++){
        const r = Math.hypot(terms[i].re, terms[i].im);
        dctx.beginPath();
        dctx.arc(chain[i].x, chain[i].y, r, 0, 2*Math.PI);
        dctx.strokeStyle = "#ccc";
        dctx.stroke();
        dctx.beginPath();
        dctx.moveTo(chain[i].x, chain[i].y);
        dctx.lineTo(chain[i+1].x, chain[i+1].y);
        dctx.strokeStyle = "red";
        dctx.stroke();
    }
    const tip = chain[chain.length -1];
    trace.push(tip);
    if(trace.length > 700){ //path lasting effect 
        trace.shift(); }
    dctx.beginPath();
    trace.forEach((p,i) => {
        if(i === 0){
            dctx.moveTo(p.x, p.y);
        }
        else{ dctx.lineTo(p.x, p.y); }
        dctx.strokeStyle = "#11e";
        dctx.lineWidth = 2;
    });
    dctx.stroke();
}

let cache = computeFourier();
function recompute(){
    cache = computeFourier();
}
function loop() {
    if(playing) t += 0.02;
    drawCircle();
    if(playing) drawRight(cache);

    requestAnimationFrame(loop);
}
loop();

// draws the circle and points on the canvas
function drawCircle(){
    ctx.clearRect(0, 0, drawCanvas.width, drawCanvas.height);
    ctx.beginPath();
    points.forEach((p,i)=>
    {
        if(i === 0){
            ctx.moveTo(p.x + CX, p.y + CY);
        }
        else{
            ctx.lineTo(p.x + CX, p.y + CY);
        }
    })
    ctx.closePath();
    ctx.strokeStyle = "black";
    ctx.stroke();

    // points
    points.forEach((p,i)=>
    {
        ctx.beginPath();
        const squareSize = 8;
        ctx.rect((p.x + CX) - squareSize/2 , (p.y + CY) - squareSize/2, squareSize,squareSize); //draw square
        ctx.fillStyle = "red";
        ctx.fill();
    });
}
drawCircle();

const rect = drawCanvas.getBoundingClientRect();

let posIndex = null;
function getPos(evt){
    const x = evt.clientX - rect.left - CX ;
    const y = evt.clientY - rect.top - CY;
    console.log(`x: ${x}, y: ${y}`);
    return {x, y };
}

drawCanvas.addEventListener('mousedown', (evt) => {
    const pos = getPos(evt);
    points.forEach((p, i) => {
        if(Math.hypot(p.x -pos.x, p.y - pos.y) < 10){
            posIndex = i;
        }
    });
});

window.addEventListener('mousemove', (evt) => {
    if(posIndex !== null){
        let pos = getPos(evt);
        points[posIndex] = pos;
        drawCircle();
        recompute();
    } 
});

window.addEventListener('mouseup', (evt) => {
    posIndex = null;
});

File: fourier.rb

# a ruby learning experience, port for ruby. 

#REQUIRES GOSU
gem install gosu
ruby fourier.rb

CX = 200
CY = 200
R  = 120

PANEL_SIZE   = 400
RIGHT_OFFSET = 420 
CONTROLS_TOP = 410

WINDOW_WIDTH  = RIGHT_OFFSET + PANEL_SIZE
WINDOW_HEIGHT = 480


# class
class FourierWindow < Gosu::Window
  def initialize
    super(WINDOW_WIDTH, WINDOW_HEIGHT)
    self.caption = "Fourier Series Drawer"

    @font = Gosu::Font.new(20)

    @n = 10 #number of draggable poitns in a circle. 
    @harmonics = @n / 2
    @points = []
    @playing = true

    @t = 0.0
    @trace = []
    @last_chain = nil
    @last_terms = nil

    @dragging = nil # :n_slider, :h_slider, point index, or nil

    init_circle
    @cache = compute_fourier

    # slider layout
    @n_slider = { x: 20,  y: CONTROLS_TOP, w: 200, min: 1, max: 20 }
    @h_slider = { x: 260, y: CONTROLS_TOP, w: 200, min: 1, max: @n }

    # buttons
    @reset_button = { x: 500, y: CONTROLS_TOP, w: 90, h: 30 }
    @pause_button = { x: 610, y: CONTROLS_TOP, w: 90, h: 30 }
  end



  def init_circle
    @points = []
    @n.times do |i| #this is a cool ruby thing. 
      angle = (i.to_f / @n) * 2 * Math::PI
      @points << { x: Math.cos(angle) * R, y: Math.sin(angle) * R }
    end
  end

  #DFT

  def compute_fourier
    n = @points.length
    coeffs = []
    n.times do |k|
      re = 0.0
      im = 0.0
      n.times do |m|
        angle = -2 * Math::PI * k * m / n.to_f
        cos = Math.cos(angle)
        sin = Math.sin(angle)
        re += @points[m][:x] * cos - @points[m][:y] * sin
        im += @points[m][:x] * sin + @points[m][:y] * cos
      end
      re /= n
      im /= n
      freq = k <= n / 2 ? k : k - n
      coeffs << { freq: freq, re: re, im: im }
    end
    coeffs.sort_by { |c| c[:freq].abs }
  end

  def recompute
    @cache = compute_fourier
  end

  def eval_at(t, terms)
    x = CX.to_f
    y = CY.to_f
    chain = [{ x: x, y: y }]
    terms.each do |term|
      angle = term[:freq] * t
      x += term[:re] * Math.cos(angle) - term[:im] * Math.sin(angle)
      y += term[:re] * Math.sin(angle) + term[:im] * Math.cos(angle)
      chain << { x: x, y: y }
    end
    chain
  end

  # update / draw 

  def update
    handle_dragging

    if @playing
      @t += 0.02
      terms = @cache.first(@harmonics)
      chain = eval_at(@t, terms)
      @last_chain = chain
      @last_terms = terms

      tip = chain.last
      @trace << tip
      @trace.shift if @trace.length > 700
    end
  end

  def draw
    draw_left_panel
    draw_right_panel
    draw_controls
  end

  def draw_left_panel
    # polygon connecting the points (closed path)
    @points.each_with_index do |p, i|
      nxt = @points[(i + 1) % @points.length]
      Gosu.draw_line(p[:x] + CX, p[:y] + CY, Gosu::Color::BLACK,
                      nxt[:x] + CX, nxt[:y] + CY, Gosu::Color::BLACK, 1)
    end

    # red squares at each point
    size = 8
    @points.each do |p|
      Gosu.draw_rect(p[:x] + CX - size / 2, p[:y] + CY - size / 2,
                      size, size, Gosu::Color::RED, 1)
    end
  end

  def draw_right_panel
    return unless @last_chain && @last_terms

    chain = @last_chain
    terms = @last_terms

    (chain.length - 1).times do |i|
      r = Math.hypot(terms[i][:re], terms[i][:im])
      draw_circle_outline(chain[i][:x] + RIGHT_OFFSET, chain[i][:y], r, Gosu::Color::GRAY)
      Gosu.draw_line(chain[i][:x] + RIGHT_OFFSET, chain[i][:y], Gosu::Color::RED,
                      chain[i + 1][:x] + RIGHT_OFFSET, chain[i + 1][:y], Gosu::Color::RED, 2)
    end

    blue = Gosu::Color::BLUE
    @trace.each_cons(2) do |a, b|
      Gosu.draw_line(a[:x] + RIGHT_OFFSET, a[:y], blue,
                      b[:x] + RIGHT_OFFSET, b[:y], blue, 2)
    end
  end

  def draw_circle_outline(cx, cy, radius, color, segments = 40)
    return if radius <= 0

    angle_step = 2 * Math::PI / segments
    segments.times do |i|
      a1 = i * angle_step
      a2 = (i + 1) * angle_step
      x1 = cx + radius * Math.cos(a1)
      y1 = cy + radius * Math.sin(a1)
      x2 = cx + radius * Math.cos(a2)
      y2 = cy + radius * Math.sin(a2)
      Gosu.draw_line(x1, y1, color, x2, y2, color, 1)
    end
  end
  def draw_controls
    # N slider
    draw_slider(@n_slider, @n)
    @font.draw_text("N: #{@n}", @n_slider[:x], @n_slider[:y] - 22, 1)

    # harmonics slider
    draw_slider(@h_slider, @harmonics)
    @font.draw_text("Harmonics: #{@harmonics}", @h_slider[:x], @h_slider[:y] - 22, 1)

    # reset button
    b = @reset_button
    Gosu.draw_rect(b[:x], b[:y], b[:w], b[:h], Gosu::Color::GRAY, 1)
    @font.draw_text("Reset", b[:x] + 15, b[:y] + 5, 2)

    # pause/play button
    b = @pause_button
    Gosu.draw_rect(b[:x], b[:y], b[:w], b[:h], Gosu::Color::GRAY, 1)
    @font.draw_text(@playing ? "Pause" : "Play", b[:x] + 15, b[:y] + 5, 2)
  end

  def draw_slider(slider, value)
    track_y = slider[:y] + 5
    Gosu.draw_rect(slider[:x], track_y, slider[:w], 4, Gosu::Color::GRAY, 1)

    frac = (value - slider[:min]).to_f / (slider[:max] - slider[:min])
    handle_x = slider[:x] + frac * slider[:w]
    Gosu.draw_rect(handle_x - 5, slider[:y] - 5, 10, 20, Gosu::Color::BLACK, 2)
  end

  #mouse interaction

  def button_down(id)
    if id == Gosu::MS_LEFT
      pos = { x: mouse_x, y: mouse_y }

      if hit?(@reset_button, pos)
        reset!
        return
      end

      if hit?(@pause_button, pos)
        @playing = !@playing
        return
      end

      if slider_hit?(@n_slider, pos)
        @dragging = :n_slider
        return
      end
      if slider_hit?(@h_slider, pos)
        @dragging = :h_slider
        return
      end
      # point dragging (left panel only)
      if pos[:x] >= 0 && pos[:x] <= PANEL_SIZE && pos[:y] >= 0 && pos[:y] <= PANEL_SIZE
        local = { x: pos[:x] - CX, y: pos[:y] - CY }
        @points.each_with_index do |p, i|
          if Math.hypot(p[:x] - local[:x], p[:y] - local[:y]) < 10
            @dragging = i
          end
        end
      end
    end
  end

  def button_up(id)
    @dragging = nil if id == Gosu::MS_LEFT
  end

  def handle_dragging
    return if @dragging.nil?

    case @dragging
    when :n_slider
      frac = ((mouse_x - @n_slider[:x]) / @n_slider[:w].to_f).clamp(0.0, 1.0)
      new_n = (@n_slider[:min] + frac * (@n_slider[:max] - @n_slider[:min])).round
      if new_n != @n
        @n = new_n
        init_circle
        @trace = []
        @h_slider[:max] = @n
        @harmonics = @n
        recompute
      end
    when :h_slider
      frac = ((mouse_x - @h_slider[:x]) / @h_slider[:w].to_f).clamp(0.0, 1.0)
      new_h = (@h_slider[:min] + frac * (@h_slider[:max] - @h_slider[:min])).round
      @harmonics = new_h if new_h != @harmonics
    when Integer
      local = { x: mouse_x - CX, y: mouse_y - CY }
      @points[@dragging] = local
      recompute
    end
  end

  def reset!
    init_circle
    recompute
  end

  def hit?(rect, pos)
    pos[:x] >= rect[:x] && pos[:x] <= rect[:x] + rect[:w] &&
      pos[:y] >= rect[:y] && pos[:y] <= rect[:y] + rect[:h]
  end

  def slider_hit?(slider, pos)
    pos[:x] >= slider[:x] - 5 && pos[:x] <= slider[:x] + slider[:w] + 5 &&
      pos[:y] >= slider[:y] - 10 && pos[:y] <= slider[:y] + 15
  end

  def needs_cursor?
    true
  end
end

FourierWindow.new.show
